lib/auth/keystore/keystore_test.go

```diff
@@ -0,0 +1,288 @@
+	testRawCert      = []byte(`-----BEGIN CERTIFICATE-----
+MIIDeTCCAmGgAwIBAgIRALmlBQhTQQiGIS/P0PwF97wwDQYJKoZIhvcNAQELBQAw
+VjEQMA4GA1UEChMHc2VydmVyMTEQMA4GA1UEAxMHc2VydmVyMTEwMC4GA1UEBRMn
+MjQ2NzY0MDEwMjczNTA2ODc3NjY1MDEyMTc3Mzg5MTkyODY5ODIwMB4XDTIxMDcx
+NDE5MDY1MloXDTMxMDcxMjE5MDY1MlowVjEQMA4GA1UEChMHc2VydmVyMTEQMA4G
+A1UEAxMHc2VydmVyMTEwMC4GA1UEBRMnMjQ2NzY0MDEwMjczNTA2ODc3NjY1MDEy
+MTc3Mzg5MTkyODY5ODIwMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
+qiD2rRJ5kq7hP55eOCM9DtdkWPMI8PBKgxaAiQ9J9YF3aNur98b8kACcTQ8ixSkH
+sLccVqRdt/Cnb7jtBSrwxJ9BN09fZEiyCvy7lwxNGBMQEaov9UU722nvuWKb+EkH
+zcVV9ie9i8wM88xpzzYO8eda8FZjHxaaoe2lkrHiiOFQRubJqHVhW+SNFQOV6OsV
+ETTZlg5rmWhA5rKiB6G0QeLHysSMJbbLMOXr1Vbu7Rqohmq0AF6EdMgix3OJz3qL
+9YDKQPAzhj7ViPzT07Pv/9vh5fjXaE5iThPT4n33uY1N2fJAnzscZvVmpxxuSOqx
+wWBqkRzIJez1vv3F+5xDLwIDAQABo0IwQDAOBgNVHQ8BAf8EBAMCAqQwDwYDVR0T
+AQH/BAUwAwEB/zAdBgNVHQ4EFgQUAJprFmwUDaYguqQJxHLBC35BeQ0wDQYJKoZI
+hvcNAQELBQADggEBAG3k42nHnJvsf3M4EPBqMFqLHJOcwt5jpfPrjLmTCAnjialq
+v0qp/JAwC3SgrDFQMNNcWTi+H+E9FqYVavysZbBd0/cFlYH9SWe9CJi5CyfbsLGD
+PX8hBRDZmmhTXMMHzyHqolVAFCf5nNQyQVeQGt3fBDh++WNjmkS+886lhag/a2hh
+nskCVzvig1/6Exp06k2mMlphC25/bNB3SDeQj+dIJdej6btZs2goQZ/5Sx/iwB5c
+xrmzqBs9YMU//QIN5ZFE+7opw5v6mbeGCCk3woH46VmVwO6mHCfLha4K/K92MMdg
+JhuTMEqUaAOZBoQLn+txjl3nu9WwTThJzlY0L4w=
+-----END CERTIFICATE-----
+`)
+	testPKCS11Key = []byte(`pkcs11:{"host_id": "server2", "key_id": "00000000-0000-0000-0000-000000000000"}`)
+
+	testRawSSHKeyPair = &types.SSHKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testRawPrivateKey,
+		PrivateKeyType: types.PrivateKeyType_RAW,
+	}
+	testRawTLSKeyPair = &types.TLSKeyPair{
+		Cert:    testRawCert,
+		Key:     testRawPrivateKey,
+		KeyType: types.PrivateKeyType_RAW,
+	}
+	testRawJWTKeyPair = &types.JWTKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testRawPrivateKey,
+		PrivateKeyType: types.PrivateKeyType_RAW,
+	}
+
+	testPKCS11SSHKeyPair = &types.SSHKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testPKCS11Key,
+		PrivateKeyType: types.PrivateKeyType_PKCS11,
+	}
+	testPKCS11TLSKeyPair = &types.TLSKeyPair{
+		Cert:    testRawCert,
+		Key:     testPKCS11Key,
+		KeyType: types.PrivateKeyType_PKCS11,
+	}
+	testPKCS11JWTKeyPair = &types.JWTKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testPKCS11Key,
+		PrivateKeyType: types.PrivateKeyType_PKCS11,
+	}
+)
+
+func TestKeyStore(t *testing.T) {
+	testcases := []struct {
+		desc       string
+		rawConfig  *keystore.RawConfig
+		shouldSkip func() bool
+		setup      func(t *testing.T)
+	}{
+		{
+			desc: "raw keystore",
+			rawConfig: &keystore.RawConfig{
+				RSAKeyPairSource: native.GenerateKeyPair,
+			},
+			shouldSkip: func() bool { return false },
+		},
+	}
+
+	for _, tc := range testcases {
+		tc := tc
```

Comment:
?

Reply:
This is a fun trick to avoid https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables

Having this be parallel makes no sense here (having a table driven test with only one case also makes no sense) but I originally wrote the whole test with the HSM testcases as it is in the next PR https://github.com/gravitational/teleport/pull/7614 and then split up the PRs. The HSMs can be a bit slow so the parallelism is nice.